### PR TITLE
Implement devices to devices reshard.

### PR DESCRIPTION
**Pull request by trivialfis**
This PR attempts to fix some multi gpu issues.

* Do full data copy when reshard device memory.
* Remove calculating row_segments for gpu_hist and gpu_sketch.

### CODE REVIEW

File: src/tree/param.h [hunk 1]
```diff
@@ -186,7 +186,7 @@ struct TrainParam : public dmlc::Parameter<TrainParam> {
         .describe("gpu to use for single gpu algorithms");
     DMLC_DECLARE_FIELD(n_gpus)
         .set_lower_bound(-1)
-        .set_default(1)
+        .set_default(-1)
```
Comment (RAMitchell) [inline src/tree/param.h]: Changing the default number of gpus is a big change with significant performance implications. Lets talk more about why you want to do this. The multi-gpu version can be slower depending on the input.

Comment (trivialfis) [reply src/tree/param.h]: OK, I will leave it unchanged.

**Comment by trivialfis** [comment]
Please don't merge until I got more confidence multi-gpu is actually smoothed.

**Comment by trivialfis** [comment]
@RAMitchell Ready for another review now. :)

**Comment by hcho3** [comment]
@trivialfis Would be nice if we can add unit tests for multi-GPU code. I'm looking into adding a slave worker with multi-GPU instance (p2.8xlarge) to the CI server.

**Comment by trivialfis** [comment]
@hcho3 Cool, I will try to make more tests for multi-gpu. :)

**Comment by trivialfis** [comment]
@hcho3 Do we split single gpu tests and multi-gpu tests, or just try to use all available gpus in the tests?

**Comment by hcho3** [comment]
@trivialfis In #3741, I am using a separate directory `python-mgpu` to indicate which tests should be run with a multi-GPU machine as supposed to a single-GPU machine.

**Pull request merged**